### DISCUSSION TITLE
My pg_statistic Is Inflated

### DISCUSSION THREAD

**Discussion by eccentrex** [supabase/supabase#22094] [Questions] [answered]
I have a project I have been running for a bit on the hobby side of things.  I have a free plan and my tables according to the database reports I run add up to around 40mb.  My usage limit is showing near max (500mb) and I couldn't figure out how.  I ran the following command as suggested online and I found my pg_statistic file is huge?  I can not figure out how to approach this.  I have restarted my project, I have manually vacuumed, I am out of ideas.  I have also listed some info on my project below if it helps give you a full idea of what I have running.  Any help sorting this issue or bringing to light why the pg_statistic is so hug would be great!  

```pgsql
select
  pg_size_pretty(pg_total_relation_size(oid)),
  relname
from
  pg_class
where
  relkind = 'r'
  and not relisshared
order by
  pg_total_relation_size(oid) desc;
```

Some info on my project:  I have a total of approximately 2 insert operations run every hour from different devices throughout the day.  I have a trigger function to adjust and disseminate the data as needed to a few different tables.  This is working well.  I also have pg_cron running two different functions on a schedule.  First function runs nightly and updates a specific row field with the cumulative data that is needed (approx. 30 values read from 1000 values, repeated for 10 rows).  Second function runs once a week and deletes data in one of the tables that is older than 2 months old to manage size.  (Vacuum is done automatically at set intervals within supabase as I understand)  Lastly I have the basic extensions installed that were there automatically when I started this project plus pg_cron, no other extentions added.  

edit:
I have reduced the size of my largest table by 60% thinking perhaps this may help.  It did not unfortunately.  

I have attempted what @GaryAustin1 suggested.  I can successfully run vacuum and analyze but vacuum full would time out and truncate gave an access error "ERROR:  42501: permission denied: "pg_statistic" is a system catalog".  I tried to increase the timeout beyond the default 2min but it appears there is a failsafe where despite my increase it is still limited to 2min higher up the chain.  

I checked a few other thigs such as total relation size vs actual data size and I get 369 total and 271 data size.  271 is still quite large.  I would be happy if I could reclaim the 98mb for the time being but ultimately as my project starts to take on more features and functions I believe I would need a real answer or fix.  

If anyone has an idea of something I could try even if its a long shot, I am willing.  If I don't find a solution here, I will try Supabase support next.  

![database size issue](https://github.com/supabase/supabase/assets/133612786/c9822140-3baa-47af-b9ca-5557f59a8619)
![database size actual](https://github.com/supabase/supabase/assets/133612786/19b072ad-677d-4e2a-97ec-a1f19cb4e93f)

**Comment by GaryAustin1** [comment]
Not an expert on  SQL at all, but that seems really out of whack for what you describe.
A quick google search seems to say pg_statistic does not get cleaned up by auto vacuum and an vacuum full is required.  

If this is not a production system probably worth a try.  If it is production maybe more research on what vacuum full will mean on such a large table as it locks records it seems.

That still leaves the question of how that table got so large.  Mine on a debug site with some cron after 2 years is 2MB (400k after vacuum full).

**Reply by eccentrex** [reply]
Thanks for the help.  I tried to do a vacuum (not full) and it went through but didn't change the size.  I tried vacuum full as pictured here, and it failed.  I might be missing something simple?  (I am doing this from the SQL editor on the supabase site)
![vacuum attempt](https://github.com/supabase/supabase/assets/133612786/f18560e3-45c5-4433-8758-882429ba672d)

**Reply by eccentrex** [reply]
sorry, my bad, here it is again (done correctly this time)
![vacuum attempt](https://github.com/supabase/supabase/assets/133612786/7e36e3a6-f18b-4cba-b7ab-ccc0ac950433)

**Reply by eccentrex** [reply]
or this attempt!
![vacuum attempt](https://github.com/supabase/supabase/assets/133612786/f7412b5e-1a85-410f-89ae-e2057a4a55ee)

**Reply by GaryAustin1** [reply]
You are timing out.
![image](https://github.com/supabase/supabase/assets/54564956/2d200f48-3c7f-4e6e-b3a6-998b4b8d56f9)
https://supabase.com/docs/guides/database/postgres/configuration#timeouts

As I cautioned though, you may need to do more research for that size table.

**Reply by eccentrex** [reply]
Thank you so much for your time, I appreciate it!  I will post my progress soon!

**Reply by eccentrex** [reply]
I have tried a few things including the following.  I still get a timeout at 2 minutes (likely a global limit to safeguard the project).  I tried to make a function to do the vacuum in portions with a limit but found I can not run vacuum within a function.  I will continue to investigate.
```
set statement_timeout to 600000; -- 10min
```

**Comment by eccentrex** [comment, marked as answer]
Rodrigo from the support team just got back to me, here was his suggestion and it worked!  It doesnt explain why it reached that size but thats fine with me.  Thanks again for the help.

"One thing that you can do to restore the pg_statistic is to pause & restore your project. This will create a logical backup, so anything in pg_catalog won't be stored in the backup, so you'll have a clean slate after restoring the project."
